zerver/lib/message.py

```diff
@@ -555,6 +555,26 @@ def bulk_access_messages(user_profile: UserProfile, messages: Sequence[Message])
             filtered_messages.append(message)
     return filtered_messages
 
+def bulk_access_messages_expect_usermessage(
+        user_profile_id: int, message_ids: Sequence[int]) -> List[int]:
+    '''
+    Like bulk_access_messages, but faster and potentially stricter.
+
+    Returns a subset of `message_ids` containing only messages the
+    user can access.  Makes O(1) database queries.
+
+    Use this function only when the user is expected to have a
+    UserMessage row for every message in `message_ids`.  If a
+    UserMessage row is missing, the message will be omitted even if
+    the user has access (e.g. because it went to a public stream.)
+
+    See also: `access_message`, `bulk_access_messages`.
+    '''
+    return UserMessage.objects.filter(
+        user_profile_id=user_profile_id,
+        message_id__in=message_ids,
+    ).values_list('message_id', flat=True)
+
```

Comment:
Just a comment on this: One corner case / race that actually happens is that the message is deleted between being sent and our trying to send the push notification event; which is why we have this block in the send code path:

```
    try:                                                                                                                                    
        (message, user_message) = access_message(user_profile, missed_message['message_id'])                                                
    except JsonableError:                                                                                                                   
        if ArchivedMessage.objects.filter(id=missed_message['message_id']).exists():                                                        
            # If the cause is a race with the message being deleted,                                                                        
            # that's normal and we have no need to log an error.                                                                            
            return                                                                                                                          
```

That race is far less likely in the remove push notifications code path, because in particular the removals are triggered by work in the UserMessage table, so if the row gets deleted.

But this does make me realize a subtle bug here: our logic for deleting messages probably needs to trigger a `remove_push_notifications` event if it deleted any with `active_mobile_push_notification` enabled.  Can be a follow-up issue.